<a href="https://colab.research.google.com/github/Suwannaket1/bsc_dpdm24/blob/main/Chapter_5_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Data

In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os
os.chdir('/content/drive/MyDrive/DPDM24')

!pwd # เช็ค path ที่กำลังทำงานอยู่

/content/drive/MyDrive/DPDM24


In [4]:
apndx_data = pd.read_csv("/content/drive/MyDrive/DPDM24/apndx.csv")

In [5]:
apndx_data

Patient_ID       Country  Age  Gender   BMI Smoking_Status  \
0       PID_000001         China   67  Female  28.6             No   
1       PID_000002  Saudi Arabia   83  Female  19.4             No   
2       PID_000003        France   52  Female  17.2             No   
3       PID_000004        Russia   71    Male  31.8            Yes   
4       PID_000005           USA   19  Female  17.7             No   
...            ...           ...  ...     ...   ...            ...   
259995  PID_259996           USA   33    Male  22.3             No   
259996  PID_259997     Australia   45  Female  27.5            Yes   
259997  PID_259998  South Africa   76    Male  27.9            Yes   
259998  PID_259999         China   68  Female  24.0             No   
259999  PID_260000        Brazil   56  Female  22.7             No   

       Alcohol_Consumption Family_History_Cancer Genetic_Mutations  \
0                      Low                   Yes                No   
1                      Low                    No                No   
2                      Low                   Yes                No   
3                      Low                    No                No   
4                      Low                    No                No   
...                    ...                   ...               ...   
259995                 Low                    No                No   
259996            Moderate                   Yes                No   
259997            Moderate                    No                No   
259998                 Low                    No                No   
259999                 Low                    No               Yes   

       Chronic_Diseases  ... Cholesterol_Level White_Blood_Cell_Count  \
0              Diabetes  ...               228                    8.1   
1              Diabetes  ...               186                    9.7   
2              Diabetes  ...               231                    8.2   
3                   NaN  ...               151                    6.8   
4          Hypertension  ...               236                    6.7   
...                 ...  ...               ...                    ...   
259995         Diabetes  ...               159                    7.5   
259996     Hypertension  ...               268                    8.0   
259997              NaN  ...               156                    7.8   
259998              NaN  ...               226                    7.9   
259999              NaN  ...               270                    6.0   

       Red_Blood_Cell_Count Platelet_Count  Tumor_Markers  Symptom_Severity  \
0                       4.4            361       Negative              Mild   
1                       6.2            235       Negative          Moderate   
2                       4.6            269       Negative          Moderate   
3                       6.0            193       Negative          Moderate   
4                       4.9            396       Negative              Mild   
...                     ...            ...            ...               ...   
259995                  5.2            344       Negative          Moderate   
259996                  5.0            195       Negative          Moderate   
259997                  5.5            201       Positive          Moderate   
259998                  5.1            386       Negative          Moderate   
259999                  4.8            339       Negative          Moderate   

        Diagnosis_Delay_Days  Treatment_Type  Survival_Years_After_Diagnosis  \
0                        691         Surgery                             6.5   
1                        275    Chemotherapy                             1.6   
2                        668    Chemotherapy                             0.7   
3                        256       Radiation                             2.1   
4                        370       Radiation                             0.5   
...                 

In [6]:
apndx_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260000 entries, 0 to 259999
Data columns (total 25 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Patient_ID                      260000 non-null  object 
 1   Country                         260000 non-null  object 
 2   Age                             260000 non-null  int64  
 3   Gender                          260000 non-null  object 
 4   BMI                             260000 non-null  float64
 5   Smoking_Status                  260000 non-null  object 
 6   Alcohol_Consumption             260000 non-null  object 
 7   Family_History_Cancer           260000 non-null  object 
 8   Genetic_Mutations               260000 non-null  object 
 9   Chronic_Diseases                129913 non-null  object 
 10  Physical_Activity_Level         260000 non-null  object 
 11  Diet_Type                       260000 non-null  object 
 12  Radiation_Exposu

In [7]:
apndx_data['Chronic_Diseases']

0             Diabetes
1             Diabetes
2             Diabetes
3                  NaN
4         Hypertension
              ...     
259995        Diabetes
259996    Hypertension
259997             NaN
259998             NaN
259999             NaN
Name: Chronic_Diseases, Length: 260000, dtype: object

In [8]:
apndx_data['Chronic_Diseases'] = apndx_data['Chronic_Diseases'].fillna('No')

In [9]:
apndx_data['Chronic_Diseases']

0             Diabetes
1             Diabetes
2             Diabetes
3                   No
4         Hypertension
              ...     
259995        Diabetes
259996    Hypertension
259997              No
259998              No
259999              No
Name: Chronic_Diseases, Length: 260000, dtype: object

In [10]:
apndx_data['Treatment_Type'] = apndx_data['Treatment_Type'].fillna('No')

In [11]:
apndx_data['Treatment_Type']

0              Surgery
1         Chemotherapy
2         Chemotherapy
3            Radiation
4            Radiation
              ...     
259995         Surgery
259996       Radiation
259997    Chemotherapy
259998       Radiation
259999         Surgery
Name: Treatment_Type, Length: 260000, dtype: object

In [12]:
set(apndx_data['Symptom_Severity'])

{'Mild', 'Moderate', 'Severe'}

In [13]:
apndx_data.groupby(['Symptom_Severity','Appendix_Cancer_Prediction']).count()

Patient_ID  Country     Age  \
Symptom_Severity Appendix_Cancer_Prediction                                
Mild             No                              110229   110229  110229   
                 Yes                              19532    19532   19532   
Moderate         No                               77436    77436   77436   
                 Yes                              13761    13761   13761   
Severe           No                               33048    33048   33048   
                 Yes                               5994     5994    5994   

                                             Gender     BMI  Smoking_Status  \
Symptom_Severity Appendix_Cancer_Prediction                                   
Mild             No                          110229  110229          110229   
                 Yes                          19532   19532           19532   
Moderate         No                           77436   77436           77436   
                 Yes                          13761   13761           13761   
Severe           No                           33048   33048           33048   
                 Yes                           5994    5994            5994   

                                             Alcohol_Consumption  \
Symptom_Severity Appendix_Cancer_Prediction                        
Mild             No                                       110229   
                 Yes                                       19532   
Moderate         No                                        77436   
                 Yes                                       13761   
Severe           No                                        33048   
                 Yes                                        5994   

                                             Family_History_Cancer  \
Symptom_Severity Appendix_Cancer_Prediction                          
Mild             No                                         110229   
                 Yes                                         19532   
Moderate         No                                          77436   
                 Yes                                         13761   
Severe           No                                          33048   
                 Yes                                          5994   

                                             Genetic_Mutations  \
Symptom_Severity Appendix_Cancer_Prediction                      
Mild             No                                     110229   
                 Yes                                     19532   
Moderate         No                                      77436   
                 Yes                                     13761   
Severe           No                                      33048   
                 Yes                                      5994   

                                             Chronic_Diseases  ...  \
Symptom_Severity Appendix_Cancer_Prediction                    ...   
Mild             No                                    110229  ...   
                 Yes                                    19532  ...   
Moderate         No                                     77436  ...   
                 Yes                                    13761  ...   
Severe           No                                     33048  ...   
                 Yes                                     5994  ...   

                                             Previous_Cancers  Blood_Pressure  \
Symptom_Severity Appendix_Cancer_Prediction                                     
Mild             No                                    110229          110229   
                 Yes                                    19532           19532   
Moderate         No                                     77436           77436   
                 Yes                                    13761           13761   
Severe           No                                     33048           33048   
                 Yes                    

In [14]:
# drop column ที่ได้หลังจากที่รู้ว่าเป็น/ไม่เป็นมะเร็งแล้ว

apndx_data = apndx_data.drop(['Survival_Years_After_Diagnosis', 'Treatment_Type', 'Diagnosis_Delay_Days'], axis=1)

In [15]:
# drop column ที่ไม่น่าจะใช้สร้าง model ได้

apndx_data = apndx_data.drop(['Patient_ID', 'Country'], axis=1)

In [16]:
apndx_data

Age  Gender   BMI Smoking_Status Alcohol_Consumption  \
0        67  Female  28.6             No                 Low   
1        83  Female  19.4             No                 Low   
2        52  Female  17.2             No                 Low   
3        71    Male  31.8            Yes                 Low   
4        19  Female  17.7             No                 Low   
...     ...     ...   ...            ...                 ...   
259995   33    Male  22.3             No                 Low   
259996   45  Female  27.5            Yes            Moderate   
259997   76    Male  27.9            Yes            Moderate   
259998   68  Female  24.0             No                 Low   
259999   56  Female  22.7             No                 Low   

       Family_History_Cancer Genetic_Mutations Chronic_Diseases  \
0                        Yes                No         Diabetes   
1                         No                No         Diabetes   
2                        Yes                No         Diabetes   
3                         No                No               No   
4                         No                No     Hypertension   
...                      ...               ...              ...   
259995                    No                No         Diabetes   
259996                   Yes                No     Hypertension   
259997                    No                No               No   
259998                    No                No               No   
259999                    No               Yes               No   

       Physical_Activity_Level       Diet_Type Radiation_Exposure  \
0                         High  Non-Vegetarian                 No   
1                          Low  Non-Vegetarian                 No   
2                          Low  Non-Vegetarian                 No   
3                         High  Non-Vegetarian                 No   
4                     Moderate  Non-Vegetarian                 No   
...                        ...             ...                ...   
259995                     Low  Non-Vegetarian                 No   
259996                     Low      Vegetarian                 No   
259997                     Low  Non-Vegetarian                 No   
259998                Moderate  Non-Vegetarian                 No   
259999                     Low      Vegetarian                 No   

       Previous_Cancers  Blood_Pressure  Cholesterol_Level  \
0                    No             156                228   
1                    No             117                186   
2                    No             115                231   
3                   Yes             149                151   
4                    No             179                236   
...                 ...             ...                ...   
259995               No             136                159   
259996              Yes             141                268   
259997               No              95                156   
259998               No             107                226   
259999              Yes              99                270   

        White_Blood_Cell_Count  Red_Blood_Cell_Count  Platelet_Count  \
0                          8.1                   4.4             361   
1                          9.7                   6.2             235   
2                          8.2                   4.6             269   
3                          6.8                   6.0             193   
4                          6.7                   4.9             396   
...                        ...                   ...             ...   
259995                     7.5                   5.2             344   
259996                     8.0                   5.0             195   
259997                     7.8                   5.5             201   
259998                     7.9                   5.1             386   
259999                     6.0                   4.8             339

# แปลง data ให้เป็นตัวเลขทั้งหมด

In [17]:
apndx_data.columns

Index(['Age', 'Gender', 'BMI', 'Smoking_Status', 'Alcohol_Consumption',
       'Family_History_Cancer', 'Genetic_Mutations', 'Chronic_Diseases',
       'Physical_Activity_Level', 'Diet_Type', 'Radiation_Exposure',
       'Previous_Cancers', 'Blood_Pressure', 'Cholesterol_Level',
       'White_Blood_Cell_Count', 'Red_Blood_Cell_Count', 'Platelet_Count',
       'Tumor_Markers', 'Symptom_Severity', 'Appendix_Cancer_Prediction'],
      dtype='object')

In [18]:
set(apndx_data['Appendix_Cancer_Prediction'])

{'No', 'Yes'}

In [19]:
# แทนตัวเลขให้กับค่าที่เรียงได้

Smoking_maping = {'Yes': 1, 'No': 0}
apndx_data['Smoking_Status'] = apndx_data['Smoking_Status'].map(Smoking_maping)

Alcohol_maping = {'Low': 0, 'Moderate': 1, 'High': 2}
apndx_data['Alcohol_Consumption'] = apndx_data['Alcohol_Consumption'].map(Alcohol_maping)

Family_maping = {'Yes': 1, 'No': 0}
apndx_data['Family_History_Cancer'] = apndx_data['Family_History_Cancer'].map(Family_maping)

Genetic_maping = {'Yes': 1, 'No': 0}
apndx_data['Genetic_Mutations'] = apndx_data['Genetic_Mutations'].map(Genetic_maping)

Physical_maping = {'Low': 0, 'Moderate': 1, 'High': 2}
apndx_data['Physical_Activity_Level'] = apndx_data['Physical_Activity_Level'].map(Physical_maping)

Radiation_Exposure_maping = {'Yes': 1, 'No': 0}
apndx_data['Radiation_Exposure'] = apndx_data['Radiation_Exposure'].map(Radiation_Exposure_maping)

Previous_maping = {'Yes': 1, 'No': 0}
apndx_data['Previous_Cancers'] = apndx_data['Previous_Cancers'].map(Previous_maping)

Tumores_maping = {'Negative': 0, 'Positive': 1}
apndx_data['Tumor_Markers'] = apndx_data['Tumor_Markers'].map(Tumores_maping)

Symptom_maping = {'Low': 0, 'Moderate': 1, 'High': 2}
apndx_data['Symptom_Severity'] = apndx_data['Symptom_Severity'].map(Symptom_maping)

apndx_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260000 entries, 0 to 259999
Data columns (total 20 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Age                         260000 non-null  int64  
 1   Gender                      260000 non-null  object 
 2   BMI                         260000 non-null  float64
 3   Smoking_Status              260000 non-null  int64  
 4   Alcohol_Consumption         260000 non-null  int64  
 5   Family_History_Cancer       260000 non-null  int64  
 6   Genetic_Mutations           260000 non-null  int64  
 7   Chronic_Diseases            260000 non-null  object 
 8   Physical_Activity_Level     260000 non-null  int64  
 9   Diet_Type                   260000 non-null  object 
 10  Radiation_Exposure          260000 non-null  int64  
 11  Previous_Cancers            260000 non-null  int64  
 12  Blood_Pressure              260000 non-null  int64  
 13  Cholesterol_Le

In [20]:
# แทนตัวเลขให้กับค่าที่เรียงไม่ได้

import pandas as pd
from sklearn.preprocessing import OneHotEncoder

# Create a OneHotEncoder object
enc = OneHotEncoder(handle_unknown='ignore')

# Fit the encoder on the categorical columns
enc.fit(apndx_data[['Chronic_Diseases', 'Gender', 'Diet_Type']])

# Transform the categorical features into one-hot encoded features
encoded_features = enc.transform(apndx_data[['Chronic_Diseases', 'Gender', 'Diet_Type']]).toarray()

# Create a new DataFrame with the encoded features
encoded_df = pd.DataFrame(encoded_features, columns=enc.get_feature_names_out(['Chronic_Diseases', 'Gender', 'Diet_Type']))

# Concatenate the encoded features with the original DataFrame
apndx_data = pd.concat([apndx_data, encoded_df], axis=1)

# Optionally, drop the original categorical columns
apndx_data = apndx_data.drop(['Chronic_Diseases', 'Gender', 'Diet_Type'], axis=1)

# Print the updated DataFrame
apndx_data.head()

Age   BMI  Smoking_Status  Alcohol_Consumption  Family_History_Cancer  \
0   67  28.6               0                    0                      1   
1   83  19.4               0                    0                      0   
2   52  17.2               0                    0                      1   
3   71  31.8               1                    0                      0   
4   19  17.7               0                    0                      0   

   Genetic_Mutations  Physical_Activity_Level  Radiation_Exposure  \
0                  0                        2                   0   
1                  0                        0                   0   
2                  0                        0                   0   
3                  0                        2                   0   
4                  0                        1                   0   

   Previous_Cancers  Blood_Pressure  ...  Appendix_Cancer_Prediction  \
0                 0             156  ...                         Yes   
1                 0             117  ...                          No   
2                 0             115  ...                          No   
3                 1             149  ...                          No   
4                 0             179  ...                          No   

   Chronic_Diseases_Diabetes  Chronic_Diseases_Hypertension  \
0                        1.0                            0.0   
1                        1.0                            0.0   
2                        1.0                            0.0   
3                        0.0                            0.0   
4                        0.0                            1.0   

   Chronic_Diseases_No  Gender_Female  Gender_Male Gender_Other  \
0                  0.0            1.0          0.0          0.0   
1                  0.0            1.0          0.0          0.0   
2                  0.0            1.0          0.0          0.0   
3                  1.0            0.0          1.0          0.0   
4                  0.0            1.0          0.0          0.0   

   Diet_Type_Non-Vegetarian  Diet_Type_Vegan  Diet_Type_Vegetarian  
0                       1.0              0.0                   0.0  
1                       1.0              0.0                   0.0  
2                       1.0              0.0                   0.0  
3                       1.0              0.0                   0.0  
4                       1.0              0.0                   0.0  

[5 rows x 26 columns]

In [21]:
Appendix_maping = {'Yes': 1, 'No': 0}
apndx_data['Appendix_Cancer_Prediction'] = apndx_data['Appendix_Cancer_Prediction'].map(Appendix_maping)
apndx_data.index

RangeIndex(start=0, stop=260000, step=1)

In [22]:
apndx_data.head()

Age   BMI  Smoking_Status  Alcohol_Consumption  Family_History_Cancer  \
0   67  28.6               0                    0                      1   
1   83  19.4               0                    0                      0   
2   52  17.2               0                    0                      1   
3   71  31.8               1                    0                      0   
4   19  17.7               0                    0                      0   

   Genetic_Mutations  Physical_Activity_Level  Radiation_Exposure  \
0                  0                        2                   0   
1                  0                        0                   0   
2                  0                        0                   0   
3                  0                        2                   0   
4                  0                        1                   0   

   Previous_Cancers  Blood_Pressure  ...  Appendix_Cancer_Prediction  \
0                 0             156  ...                           1   
1                 0             117  ...                           0   
2                 0             115  ...                           0   
3                 1             149  ...                           0   
4                 0             179  ...                           0   

   Chronic_Diseases_Diabetes  Chronic_Diseases_Hypertension  \
0                        1.0                            0.0   
1                        1.0                            0.0   
2                        1.0                            0.0   
3                        0.0                            0.0   
4                        0.0                            1.0   

   Chronic_Diseases_No  Gender_Female  Gender_Male  Gender_Other  \
0                  0.0            1.0          0.0           0.0   
1                  0.0            1.0          0.0           0.0   
2                  0.0            1.0          0.0           0.0   
3                  1.0            0.0          1.0           0.0   
4                  0.0            1.0          0.0           0.0   

   Diet_Type_Non-Vegetarian  Diet_Type_Vegan  Diet_Type_Vegetarian  
0                       1.0              0.0                   0.0  
1                       1.0              0.0                   0.0  
2                       1.0              0.0                   0.0  
3                       1.0              0.0                   0.0  
4                       1.0              0.0                   0.0  

[5 rows x 26 columns]

# Classification

## แบ่ง Train กับ Test

### เก็บ Test เอาไว้ วัดผลตอนจบ

In [23]:
from sklearn.model_selection import train_test_split

# Assuming apndx_data is your DataFrame
train_data, test_data = train_test_split(apndx_data, train_size=200000, random_state=6) # 20% for testing

print("Training data shape; ", train_data.shape)
print("Testing data shape; ", test_data.shape)

Training data shape;  (200000, 26)
Testing data shape;  (60000, 26)


In [24]:
train_data.groupby(['Appendix_Cancer_Prediction']).count()

Age     BMI  Smoking_Status  \
Appendix_Cancer_Prediction                                   
0                           169786  169786          169786   
1                            30214   30214           30214   

                            Alcohol_Consumption  Family_History_Cancer  \
Appendix_Cancer_Prediction                                               
0                                        169786                 169786   
1                                         30214                  30214   

                            Genetic_Mutations  Physical_Activity_Level  \
Appendix_Cancer_Prediction                                               
0                                      169786                   169786   
1                                       30214                    30214   

                            Radiation_Exposure  Previous_Cancers  \
Appendix_Cancer_Prediction                                         
0                                       169786            169786   
1                                        30214             30214   

                            Blood_Pressure  ...  Symptom_Severity  \
Appendix_Cancer_Prediction                  ...                     
0                                   169786  ...             59732   
1                                    30214  ...             10612   

                            Chronic_Diseases_Diabetes  \
Appendix_Cancer_Prediction                              
0                                              169786   
1                                               30214   

                            Chronic_Diseases_Hypertension  \
Appendix_Cancer_Prediction                                  
0                                                  169786   
1                                                   30214   

                            Chronic_Diseases_No  Gender_Female  Gender_Male  \
Appendix_Cancer_Prediction                                                    
0                                        169786         169786       169786   
1                                         30214          30214        30214   

                            Gender_Other  Diet_Type_Non-Vegetarian  \
Appendix_Cancer_Prediction                                           
0                                 169786                    169786   
1                                  30214                     30214   

                            Diet_Type_Vegan  Diet_Type_Vegetarian  
Appendix_Cancer_Prediction                                         
0                                    169786                169786  
1                                     30214                 30214  

[2 rows x 25 columns]

In [25]:
test_data.groupby(['Appendix_Cancer_Prediction']).count()

Age    BMI  Smoking_Status  Alcohol_Consumption  \
Appendix_Cancer_Prediction                                                      
0                           50927  50927           50927                50927   
1                            9073   9073            9073                 9073   

                            Family_History_Cancer  Genetic_Mutations  \
Appendix_Cancer_Prediction                                             
0                                           50927              50927   
1                                            9073               9073   

                            Physical_Activity_Level  Radiation_Exposure  \
Appendix_Cancer_Prediction                                                
0                                             50927               50927   
1                                              9073                9073   

                            Previous_Cancers  Blood_Pressure  ...  \
Appendix_Cancer_Prediction                                    ...   
0                                      50927           50927  ...   
1                                       9073            9073  ...   

                            Symptom_Severity  Chronic_Diseases_Diabetes  \
Appendix_Cancer_Prediction                                                
0                                      17704                      50927   
1                                       3149                       9073   

                            Chronic_Diseases_Hypertension  \
Appendix_Cancer_Prediction                                  
0                                                   50927   
1                                                    9073   

                            Chronic_Diseases_No  Gender_Female  Gender_Male  \
Appendix_Cancer_Prediction                                                    
0                                         50927          50927        50927   
1                                          9073           9073         9073   

                            Gender_Other  Diet_Type_Non-Vegetarian  \
Appendix_Cancer_Prediction                                           
0                                  50927                     50927   
1                                   9073                      9073   

                            Diet_Type_Vegan  Diet_Type_Vegetarian  
Appendix_Cancer_Prediction                                         
0                                     50927                 50927  
1                                      9073                  9073  

[2 rows x 25 columns]

In [26]:
9073/50927

0.17815696978027373

## เลือก Model ที่ดีที่สุด (Model Selection)


1.   Decision Tree
2.   (Short) Decision Tree
3.   Naive Bayes



## Split data into 10 folds

In [27]:
from sklearn.model_selection import KFold, cross_val_score
kf = KFold(n_splits=10, shuffle=True, random_state=6)

In [28]:
X_train = train_data.drop('Appendix_Cancer_Prediction', axis=1) # X ตัวใหญ่เพราะเป็นเวกเตอร์
y_train = train_data['Appendix_Cancer_Prediction']

## Decision Tree

In [29]:
# 1.import
from sklearn.tree import DecisionTreeClassifier

# 2.define
dt_classifier = DecisionTreeClassifier() #เป็น Default

# 3.train and # 4.evaluation (รวบ 2 ขั้นนี้ เพราะเป็น Cross Validation)
cv_scores = cross_val_score(dt_classifier, X_train, y_train, cv=kf, scoring='accuracy')

In [30]:
# Print
print("Cross-validation scores:", cv_scores) # score ของแต่ละก้อน
print("Mean accuracy:", np.mean(cv_scores)) # แล้วเอา score มาหา mean
print("Standard deviation:", np.std(cv_scores))

Cross-validation scores: [0.72205 0.7258  0.72495 0.72465 0.72135 0.72665 0.7265  0.724   0.7284
 0.72445]
Mean accuracy: 0.72488
Standard deviation: 0.0020119890655766583


มีข้อสอบ 100 ข้อ ทำถูก 72 ข้อ

  ## (Shport) Decision

In [31]:
# 1.import
from sklearn.tree import DecisionTreeClassifier

# 2.define
dt_classifier2 = DecisionTreeClassifier(max_depth=5) #เป็น Default

# 3.train and # 4.evaluation (รวบ 2 ขั้นนี้ เพราะเป็น Cross Validation)
cv_scores = cross_val_score(dt_classifier2, X_train, y_train, cv=kf, scoring='accuracy')

In [32]:
# Print
print("Cross-validation scores:", cv_scores) # score ของแต่ละก้อน
print("Mean accuracy:", np.mean(cv_scores)) # แล้วเอา score มาหา mean
print("Standard deviation:", np.std(cv_scores))

Cross-validation scores: [0.8491  0.84865 0.84905 0.84875 0.849   0.8506  0.84955 0.8494  0.8453
 0.8493 ]
Mean accuracy: 0.84887
Standard deviation: 0.001297343439494716


## Naive Bayes

In [33]:
#import
from sklearn.naive_bayes import GaussianNB
#define
nb_classifier = GaussianNB()
#cross
cv_scores3 = cross_val_score(nb_classifier, X_train, y_train, cv=kf, scoring='accuracy')

ValueError: 
All the 10 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
10 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.11/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/sklearn/naive_bayes.py", line 266, in fit
    return self._partial_fit(
           ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/sklearn/naive_bayes.py", line 426, in _partial_fit
    X, y = validate_data(self, X, y, reset=first_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py", line 2961, in validate_data
    X, y = check_X_y(X, y, **check_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py", line 1370, in check_X_y
    X = check_array(
        ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py", line 1107, in check_array
    _assert_all_finite(
  File "/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py", line 120, in _assert_all_finite
    _assert_all_finite_element_wise(
  File "/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py", line 169, in _assert_all_finite_element_wise
    raise ValueError(msg_err)
ValueError: Input X contains NaN.
GaussianNB does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values


In [ ]:
# Print cross-validation results
print("Cross-validation scores:", cv_scores3)
print("Average cross-validation score:", cv_scores3.mean())
print("Standard deviation of cross-validation scores:", cv_scores3.std())

## เลือก Short Dicision Tree มา train ด้วย all training data

In [ ]:
#import
from sklearn.tree import DecisionTreeClassifier
#define
dt_classifier2 = DecisionTreeClassifier(max_depth=5)

dt_classifier2 = dt_classifier2.fit(X_train, y_train)

In [ ]:
# prompt: plot the trained

import matplotlib.pyplot as plt
from sklearn.tree import plot_tree

# Assuming dt_classifier2 is your trained DecisionTreeClassifier
plt.figure(figsize=(50,25))
plot_tree(dt_classifier2, filled=True, feature_names=X_train.columns, class_names=['0','1'])
plt.show()

# Evaluate with Test data

In [ ]:
X_test = test_data.drop('Appendix_Cancer_Prediction', axis = 1)
y_actual = test_data['Appendix_Cancer_Prediction']

In [ ]:
# Predict
y_pred = dt_classifier2.predict(X_test)

In [ ]:
# Evaluate
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_actual, y_pred)
print(f"Accuracy on test data: {accuracy}")

# Quiz 12 หาต้นไม้ตัดสินใจที่มี average cross-validation accuracy มากกว่า 84.887%

In [35]:
#import
from sklearn.tree import DecisionTreeClassifier
#define
dt_classifierq = DecisionTreeClassifier(max_depth=1, random_state=6)

#cross
cv_scoresq = cross_val_score(dt_classifierq, X_train, y_train, cv=kf, scoring='accuracy')
# Print cross-validation results
print("Cross-validation scores:", cv_scoresq)
print("Average cross-validation score:", cv_scoresq.mean())
print("Standard deviation of cross-validation scores:", cv_scores.std())

Cross-validation scores: [0.84915 0.84865 0.8491  0.8488  0.84905 0.8506  0.84955 0.8496  0.8454
 0.8494 ]
Average cross-validation score: 0.8489299999999999
Standard deviation of cross-validation scores: 0.001297343439494716
